lib/versions.ps1

```diff
@@ -1,53 +1,210 @@
-            $ver_a[$i] = "$($ver_a[$i])"
+    if ($splitReferenceVersion[0] -eq 'nightly' -and $splitDifferenceVersion[0] -eq 'nightly') {
+        return 0
+    }
+
+    for ($i = 0; $i -lt [Math]::Max($splitReferenceVersion.Length, $splitDifferenceVersion.Length); $i++) {
+        if ($i -ge $splitReferenceVersion.Length) {
+            if ($splitDifferenceVersion[$i] -match "alpha|beta|rc|pre") {
+                return -1
+            } else {
+                return 1
+            }
+        }
+        if ($i -ge $splitDifferenceVersion.Length) {
+            if ($splitReferenceVersion[$i] -match "alpha|beta|rc|pre") {
+                return 1
+            } else {
+                return -1
+            }
+        }
+
+        if (($splitReferenceVersion[$i] -match "\.") -or ($splitDifferenceVersion[$i] -match "\.")) {
+            $Result = Compare-Version $splitReferenceVersion[$i] $splitDifferenceVersion[$i] -Delimiter '\.'
+            if ($Result -ne 0) {
+                return $Result
+            } else {
+                continue
+            }
         }
 
-        if($ver_a[$i] -gt $ver_b[$i]) { return 1; }
-        if($ver_a[$i] -lt $ver_b[$i]) { return -1; }
+        if ($null -ne $splitReferenceVersion[$i] -and $null -ne $splitDifferenceVersion[$i]) {
+            # don't try to compare int to string
+            if ($splitReferenceVersion[$i] -is [string] -and $splitDifferenceVersion[$i] -isnot [string]) {
+                $splitDifferenceVersion[$i] = "$($splitDifferenceVersion[$i])"
+            }
+            if ($splitDifferenceVersion[$i] -is [string] -and $splitReferenceVersion[$i] -isnot [string]) {
+                $splitReferenceVersion[$i] = "$($splitReferenceVersion[$i])"
+            }
+        }
+
+        if ($splitDifferenceVersion[$i] -gt $splitReferenceVersion[$i]) { return 1 }
+        if ($splitDifferenceVersion[$i] -lt $splitReferenceVersion[$i]) { return -1 }
     }
-    if($ver_b.length -gt $ver_a.length) { return -1 }
+
     return 0
 }
 
+# Deprecated
+# Not used anymore in scoop core
 function qsort($ary, $fn) {
+    warn '"qsort" is deprecated. Please avoid using it anymore.'
     if($null -eq $ary) { return @() }
     if(!($ary -is [array])) { return @($ary) }
 
     $pivot = $ary[0]
     $rem = $ary[1..($ary.length-1)]
 
-    $lesser = qsort ($rem | Where-Object { (& $fn $_ $pivot) -lt 0 }) $fn
+    $lesser = qsort ($rem | Where-Object { (& $fn $pivot $_) -lt 0 }) $fn
 
-    $greater = qsort ($rem | Where-Object { (& $fn $_ $pivot) -ge 0 }) $fn
+    $greater = qsort ($rem | Where-Object { (& $fn $pivot $_) -ge 0 }) $fn
 
     return @() + $lesser + @($pivot) + $greater
 }
-function sort_versions($versions) { qsort $versions compare_versions }
+
+# Deprecated
+# Not used anymore in scoop core
+function sort_versions($versions) {
+    warn '"sort_versions" is deprecated. Please avoid using it anymore.'
+    qsort $versions Compare-Version
+}
```

Comment:
```suggestion
}

function compare_versions($a, $b) {
    Show-DeprecatedWarning $MyInvocation 'Compare-Version'
    return Compare-Version -ReferenceVersion $a -DifferenceVersion $b
}

function versions($app, $global) {
    Show-DeprecatedWarning $MyInvocation 'Get-InstalledVersion'
    return Get-InstalledVersion -App $app -Global $global
}

function latest_version($app, $bucket, $url) {
    Show-DeprecatedWarning $MyInvocation 'Get-LatestVersion'
    return Get-LatestVersion -App $app -Bucket $bucket -URL $url
}
```

Reply:
I don't think anyone will use these functions elsewhere... But we might deprecate every unused function.

Reply:
You need to deprecate all functions. There are tools, which are utilizing scoop core. Also there is chance of missing some reference to this function.

Reply:
You need to deprecate all functions. There are tools, which are utilizing scoop core. Also there is chance of missing some reference to this function.